src/transformers/generation/configuration_utils.py

```diff
@@ -330,6 +330,18 @@ class GenerationConfig(PushToHubMixin):
         return_legacy_cache (`bool`, *optional*, default to `True`):
             Whether to return the legacy or new format of the cache when `DynamicCache` is used by default.
 
+        > Parameters specific to vision-language generation models such as [Chameleon](https://arxiv.org/abs/2405.09818v1)
+
+        multimodal_generation_mode (`Literal["text-only", "image-only", "interleaved-text-image", "free"]`, *optional*, defaults to `None`):
+            Chameleon can generate text, images, or both in an interleaved manner. However, only text generation is
+            supported by the official model checkpoint. This flag enables the other modes for use with finetuned versions
+            of the model such as [Anole](https://arxiv.org/abs/2407.06135).
+            - If set to `"text-only"`, logits for image tokens will be masked out during generation.
+            - If set to `"image-only"`, logits for non-image tokens will be masked out during generation.
+            - If set to `"free"`, the logits are left as-is.
+            - For `"interleaved-text-image"`, Chameleon implements a finite state machine to dynamically switch between text and image modalities.
+                This library does not support this mode yet.
+
```

Comment:
Sorry if I wasn't clear, I meant Chameleon's generation config that is on the hub. Adding args to the general generation config is not a good idea, if it's going to be used only by one model.

Let's see how we can make `generate()` happy. We can:
- When saving the model, add a field `model.generation_config="text-only"` by setting the default to text mode and saving it on the hub
- Move Chameleon logits processor to the other processors, more comments below
- Add a `generate()` method in ConditionalGeneration module that does model-specific preparation, in our case takes generation mode and makes a LogitsProcessor out of it. Then calls `super().generate()` with all kwargs
- Optionally, take generation output and run `decode_tokens` if image mode. In case you can make an example with external library for FSM and interleaved model, do separation of image from text here. And return a custom GenerationDecoderOnlyOutput, which will have an extra field for `pixel values`

Usually I would find a way w/o custom generate, but Chameleon can be an exception given that it's the only model that generates images. Also, custom generate in this way is less prone to bugs from refactoring general `generate()`, as we only prepare and pass a model-specific processor.

Reply:
Hi @zucchini-nlp!

I've:

1. Implemented a new generation config for Chameleon
2. Modified the utils to support custom generation config classes
3. Added a custom `generate` func to `ChameleonForConditionalGeneration`

```python
class ChameleonGenerationConfig(GenerationConfig):
    """Generation Config for [Chameleon](https://huggingface.co/docs/transformers/model_doc/chameleon)

    Args:
        multimodal_generation_mode (`Literal["text-only", "image-only", "interleaved-text-image", "unrestricted"]`, *optional*, defaults to `None`):
            Chameleon can generate text, images, or both in an interleaved manner. However, only text generation is
            supported by the official model checkpoint. This flag enables the other modes for use with finetuned versions
            of the model such as [Anole](https://arxiv.org/abs/2407.06135).
            - If set to `"unrestricted"`, the logits are left as-is.
            - If set to `"text-only"`, logits for image tokens will be masked out during generation.
            - If set to `"image-only"`, logits for non-image tokens will be masked out during generation.
            - For `"interleaved-text-image"`, Chameleon implements a finite state machine to dynamically switch between text and image modalities.
                Here, we simply use logits processors that exclusively allow image tokens to be generated within a relative window after the
                begin image token and disallow them elsewhere.
    """

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.multimodal_generation_mode = kwargs.pop("multimodal_generation_mode", "text-only")
```

Reply:
`multimodal_generation_mode` is now also in the converted model here: https://huggingface.co/leloy/Anole-7b-v0.1-hf/blob/main/generation_config.json

![image](https://github.com/user-attachments/assets/7ead0ae4-4fdf-4c14-8d57-83f3ef1076b1)